### PR TITLE
gh-264 Base changes to support theming of application

### LINKED ISSUE #264: Add Theming to Mdm Explorer: Groundwork
# Description

The Mdm Explorer needs to be themed dynamically, so that it does not have to be rebuilt when a customer wishes to change the appearance of the application. Initial research was done in https://github.com/MauroDataMapper/mdm-ui/issues/694 about how to achieve this in Angular. This task is to implement the groundwork to allow for themes to be dynamically set.

# Out of Scope
In this task, the theme source will be hard coded. Another task will implement the changes to extract a theme source from the database, and apply it.

# Technical Detail

The following is quite detailed, but is by no means set in stone. The developer should use their common sense and adapt the plan as necessary to achieve the goals of this task.

1. Rename default-theme to active-theme.

2. Rename `src/styles/themes/_default.scss` to `src/styles/themes/_active.scss`

3. In `src/styles/themes/_active.scss`, rename all `$default` prefixes to be `$active`.

4. In `src/styles/themes/_active.scss`, rename the colours in the theme to be variables as based on https://medium.com/angular-in-depth/build-truly-dynamic-theme-with-css-variables-539516e95837 e.g:
```
$active-palette-primary: (
  50 : var(--theme-primary-50),
  100 : var(--theme-primary-100),
...
  A400 : var(--theme-primary-A400),
  A700 : var(--theme-primary-A700),
  contrast: (
    50: var(--theme-primary-contrast-50),
    100: var(--theme-primary-contrast-100),
...
    A400: var(--theme-primary-contrast-A400),
    A700: var(--theme-primary-contrast-A700),
  )
); 
```
You will need to do this for all three palette themes.

5. In `src/styles/themes/_active.scss`, rename the font items to be variables:
```
$default-typography: mat-typography-config(
  $font-family: "var(--theme-font-family)",
  $body-1: mat-typography-level(var(--theme-typography-body-1)),
  $body-2: mat-typography-level(var(--theme-typography-body-2)),
  $headline: mat-typography-level(var(--theme-typography-headline)),
  $title: mat-typography-level(var(--theme-typography-title)),
  $subheading-2: mat-typography-level(var(--theme-typography-subheading-2)),
  $subheading-1: mat-typography-level(var(--theme-typography-subheading-1)),
  $button: mat-typography-level(var(--theme-typography-button)),
  $input: mat-typography-level(var(--theme-typography-input)),
);
```

6. Change `src/app/app.component.html` reference `active-theme` instead of `default-theme`.

7. Search for any other references to `default-theme` and change them to `active-theme`.

8. Next we need to replace named colours with more versatile names.
Find and replace all references to `$color-mauro-dark-green` with `$color-primary`
Find and replace all references to `$color-mauro-gold` with `$color-accent`
Find and replace all references to `$color-mauro-light-gold` with `$color-light-accent`
Find and replace all references to `$color-dark-navy` with `$color-hyperlink`
Find and replace all references to `$color-white` with `$color-background`
Find and replace all references to `$color-gray-75` with `$color-row-header-background`

`src/styles/base/_colours.scss`
Set:
```
$color-background: var(--theme-color-background)
$color-primary: var(--theme-color-primary)
$color-accent: var(--theme-color-accent)
$color-light-accent: var(--theme-color-light-accent)
$color-warn: var(--theme-color-warn)
$color-hyperlink: var(--theme-color-hyperlink)
$color-row-header-background: var(--theme-color-row-header-background)
```

Add new fields:
```
$chip-unsent-color: var(--theme-color-chip-unsent)
$chip-submitted-color: var(--theme-color-chip-submitted)
$chip-text-color-unsent: var(--theme-color-chip-unsent-text) 
$chip-text-color-submitted: var(--theme-color-chip-submitted-text) 
```

Rename fields:
`$color-yellow` to `$request-count-color` and set to `var(--theme-color-request-count) `
`$color-gray-85` to `$color-description-info-underline` and set to `var(--theme-color-description-info-underline)`
Then delete the second `$color-description-info-underline` at the bottom.
`$color-black` to `$color-stroked-border` and set to `var(--theme-color-stroked-border)`
Then delete the second `$color-stroked-border` at the bottom.

Delete fields:
`$color-gray-90`
`$color-gray-38`

`src/app/data-explorer/request-status-chip/request-status-chip.component.scss`
Delete lines: 
```
$chip-unsent-color: $color-teal;
$chip-submitted-color: $color-light-blue;
$chip-text-color-light: $color-white;
```

Replace `$chip-text-color-light` references with `$chip-text-color-unsent` and `$chip-text-color-submitted` as appropriate.

`src/app/shared/header/header.component.scss`
Delete `$request-count-color`.

9. Create a new service called `ThemeService`. We will add theme functions to this service. For now we need a function that will just return hard coded default values. A subsequent task will update the innards of this function to make the required api calls to extract the theme data from the backend. Call the function `loadTheme` and return a type object that holds all the theme variables.  The default variables need to be assigned as follows.
```
theme-color-primary: #19381f
theme-color-accent: #cdb980
theme-color-light-accent: #f3eee0
theme-color-warn: #a5122a
theme-color-chip-unsent: #3dbaae
theme-color-chip-submitted: #008bce
theme-color-chip-unsent-text: #fff
theme-color-chip-submitted-text: #fff
theme-color-hyperlink: #003752
theme-color-request-count: #ffe603
theme-color-description-info-underline: #d7d7d7
theme-color-item-label-hover: #000
theme-color-stroked-border: #000
theme-color-row-header-background: #c4c4c4
theme-font-family: "Nunito Sans, sans-serif",
theme-typography-body-1: "17px, 26px, 500"
theme-typography-body-2: "18px, 26px, 500"
theme-typography-headline: "36px, 42px, 600"
theme-typography-title: "32px, 38px, 600"
theme-typography-subheading-2: "26px, 32px, 600"
theme-typography-subheading-1: "22px, 28px, 600"
theme-typography-button: "20px, 26px, 600"
theme-typography-input-typography-level: "18px, 1.125, 500"
```

10. We need to generate the three palettes. This is the complex bit. Have a look at the following code. https://github.com/mbitson/mcg/blob/master/scripts/controllers/ColorGeneratorCtrl.js  We need to convert the relevant sections into typescript and make sure the MIT license header is included. What we are aiming for is the code that generates the different shades of colour based on the base colour. Again, have a read of https://medium.com/angular-in-depth/build-truly-dynamic-theme-with-css-variables-539516e95837 and the end game should be clear. The range of colours should be saved in a palette type object. The base color of each pallette is assigned to the 500 variable. Use `$color-primary`, `$color-accent` and `$color-warn` as the base colours for each of the three palettes.

11. The "loaded" variables then need to be applied to the website using another function added to the `ThemeService` called `applyTheme`, which takes a data type holding the loaded variables as a parameter. This can apply the variables to the web page with lines like the following:
`document.documentElement.style.setProperty('--theme-primary-A400', theme.colors.primary.A400);` 
The `loadTheme` and `applyTheme` functions should be called in `/src/app/app.component.ts` at the start of `ngOnInit`.

After all this is done the website should appear exactly as it did before the changes. 

# Testing

The tester should change the hardcoded colours in the `loadTheme` function, rebuild their local copy, and see if the colours change as expected.

### PR DESCRIPTION
**Pull request by pjmonks**
- Rename existing Angular Material theme from "default-theme" to "active-theme" - files, classes and references
- Create ThemeService to manage loading and applying themes to the DOM. Loading is hardcoded for now using existing colour scheme
- Generate Material palette colours for application using only 3 base colours
- Fix typography settings for theme and make customisable
- Update SCSS stylesheet to use latest Angular Material mixins
- Add back the missing typography settings to the theme, default to the existing values
- Rename existing Sass colour variables
- Add colour references to components that are not Angular Material
- Fix _resets.scss because Sass and CSS3 variables were incompatible. Darken colour value using HSL instead
- Change colour of data-class-row component to fit with the default colour scheme better
- Add request status colours to theme
- Add request counter theme colour
- Tests to ensure that CSS variables are set correctly
- Create documentation page for theming

**Note:** requires a new dependency on the `tinycolor2` library, please run `npm install` before testing.

Resolves #264 

# Screenshots

Default, existing theme:

image

Testing modifying the primary colour by temporarily altering the `defaultTheme.material.colors.primary` field:

image

### CODE REVIEW

**Comment by jamesrwelch** [comment]
Looking good 👍

**Comment by jamesrwelch** [comment]
(That wasn't an official review)

**Comment by NigelPalmer** [comment]
@pjmonks , I can't get this to build. I keep getting this error:

Error: src/app/shared/theme.service.ts:21:23 - error TS2307: Cannot find module 'tinycolor2' or its corresponding type declarations.

21 import tinycolor from 'tinycolor2';

**Comment by pjmonks** [comment]
> @pjmonks , I can't get this to build. I keep getting this error:
> 
> Error: src/app/shared/theme.service.ts:21:23 - error TS2307: Cannot find module 'tinycolor2' or its corresponding type declarations.
> 
> 21 import tinycolor from 'tinycolor2';

I've updated the PR description to state you need to run `npm install` first.

**Review by NigelPalmer — CHANGES_REQUESTED** [review]
Overall this is looking good and it allows for colours to be easily changed. But, there are a few problems.

### Default Theme Issues
The default colours that are now being used are not the same as they were before. There are a number of discrepancies between the old theming and the new theming. I only checked the Green, but I expect the gold and red schemes to also have discrepancies. The left column are the original colours, and the right column are the new calculated colours.
Pallette Differences

The colour of the class header has changed from grey to gold:

Original:
classColourDev

New:
classColourThemed

The progress bar has changed from Green to Grey.

Original:
NoElementsDev

New:
NoElementsThemed

The unsent badge now has black text. (although this is probably the correct contrast colour)
unsentColorChanged


### Custom Theme Issues
The $color-white is used in a lot of css files still. I think I overlooked this when specifying the task. The knock on effect is that text is not being contrasted properly if a light custom colour is chosen. To improve the theming we either have to choose the colour automatically or allow the user to set it.
whiteText

File: src/app/shared/header/header.component.scss [hunk 1]
```diff
@@ -23,8 +23,8 @@ $link-color: #ffffff;
 $screen-size-md: 1000px;
 $screen-size-sm: 620px;
 $profile-img-width: 32px;
-$header-background: $color-mauro-dark-green;
-$request-count-color: $color-yellow;
+$header-background: $color-primary;
+$request-count-color: $color-requestCount; //$color-yellow;
```
[the remark this answers is not in the document]
Comment (pjmonks) [reply src/app/shared/header/header.component.scss]: Removed

File: src/styles/base/_colours.scss [hunk 2]
```diff
@@ -27,23 +27,53 @@ $color-gray-90: #e6e6e6;
 $color-gray-85: #d7d7d7;
 $color-gray-75: #c4c4c4;
 $color-gray-38: #606060;
```
[the remark this answers is not in the document]
Comment (pjmonks) [reply src/styles/base/_colours.scss]: Removed the gray values which are not used anymore

**Comment by pjmonks** [comment]
Updated with changes requested from feedback:

- Code cleanup
- Unsent and submitted badge colours have been adjusted
- Fixed Material colour palette generation, was using wrong algorithm before. This also fixes the progress bar colours
- Updated other components to have correct colour contrast of text when a theme colour is used as a background.
- Added new theme colour "classRow" to put back the grey background for `data-class-row` component

Note that to support wider theming there may be additional minor tweaks to the UI.

**Review by NigelPalmer — APPROVED** [review]
This all looks good now. All my complaints have been dealt with and it is easy to change the colours of the website. A good job, well done! :thumbsup:

**Pull request merged**